rfc/20250211-s3-locking-with-conditional-writes.md

```diff
@@ -0,0 +1,133 @@
+
+#### Migration paths
+##### I have no locking enabled
+In this case, the user can just add the new `use_lockfile=true` and run `tofu init -reconfigure`.
+
+##### I have DynamoDB locking enabled
+In case the user is having DynamoDB enabled, there are two paths forward:
+1. Add the new attribute `use_lockfile=true` and run `tofu init -reconfigure`
+   * Later, remove the `dynamodb_table` attribute and run `tofu init -reconfigure` again
+2. Add the new attribute `use_lockfile=true`, remove the `dynamodb_table` one and run `tofu init -reconfigure`
+
+OpenTofu recommends to have both locking mechanisms enabled for a limited amount of time and later remove the DynamoDB locking.
+### Technical Approach
+
+In order to achieve and ensure a proper state locking via S3 bucket, we want to attempt to create the locking object only when it is missing. 
+In order to do so we need to call `s3client.PutObject` with the property `IfNoneMatch: "*"`.
+For more information, please check the [official documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-writes.html#conditional-write-key-names).
+
+But the simplified implementation would look like this:
+```go
+input := &s3.PutObjectInput{
+    Bucket:            aws.String(bucket),
+    Key:               aws.String(key),
+    Body:              bytes.NewReader([]byte(lockInfo)),
+    IfNoneMatch:       aws.String("*"),
+}
+_, err := actor.S3Client.PutObject(ctx, input)
+```
+
+The `err` returned above should be handled accordingly with the [behaviour defined](https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-writes.html) in the official docs:
+* HTTP 200 (OK) - Means that the locking object was not existing. Therefore, the lock can be considered as acquired.
+  * For buckets with versioning enabled, if there's no current object version with the same name, or if the current object version is a delete marker, the write operation succeeds.
+* HTTP 412 (Precondition Failed) - Means that the locking object is already there. Therefore, the whole process should exit because the lock couldn't be acquired.
+* HTTP 409 (Conflict) - Means that the there was a conflict of concurrent requests. AWS recommends to retry the request in such cases, but we could just handle this similarly to the 412 case.
+
+#### Digest updates
+> [!NOTE]
+> Right now, when locking is enabled on DynamoDB, at the moment of updating the state object content, OpenTofu also writes an entry in DynamoDB with the MD5 sum of the state object.
+> The reason is to be able to check the integrity of the state object from the S3 bucket in a future run. This is done by reading the digest from DynamoDB and comparing it with the ETag attribute of the state object from S3. 
+
+By moving to the S3 based locking, OpenTofu will hold no other file for the digest of the state object, since the digest is kept in `ETag` header of the object.
```

Comment:
> since the digest is kept in `ETag` header of the object

I believe digest check was introduced to tackle the eventual consistency nature of s3 before 2020, so we can clearly state this is deprecated behavior and we don't need that at all. Unless, we also want to check if that object hasn't been updated, but that part we handle via a separate lock file. So I am not sure if checksum validation is something we want to keep.

WDYT?

Reply:
Reworked in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179.
Yes, this is the idea. Once only the S3 locking is in use, the digest of the state will not be handled anymore by OpenTofu.